<a href="https://colab.research.google.com/github/Brandyli/Python_Practice/blob/master/Using_the_Pandas_Python_Library.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

analyze NBA results provided by FiveThirtyEight in a 17MB CSV file. 
Create a script download_nba_all_elo.py to download the data

When I excute the script, it will save the file nba_all_elo.csv in your current working directory.

In [37]:
import requests

download_url = "https://raw.githubusercontent.com/fivethirtyeight/data/master/nba-elo/nbaallelo.csv"
target_csv_path = "nba_all_elo.csv"

response = requests.get(download_url)
response.raise_for_status()    # Check that the request was successful
with open(target_csv_path, "wb") as f:
    f.write(response.content)
print("Download ready.")

Download ready.


I use pd.read_csv to read my dataset and store it as a DataFrame object. similar functions like read_json(), read_html(), and read_sql_table(). 


In [38]:
import pandas as pd
nba = pd.read_csv("nba_all_elo.csv")
type(nba)

pandas.core.frame.DataFrame

len() to determine the number of row
.shape to see its dimensionality or column

In [39]:
len(nba)
nba.shape

(126314, 23)

In [0]:
pd.set_option("display.max.columns", None)

In [42]:
nba.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts    elo_i    elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.00  1293.28      40.29    NYK   
1            0     NYK   Knicks   68  1300.00  1306.72      41.71    TRH   
2            0     CHS    Stags   63  1300.00  1309.65      42.01    NYK   
3            0     NYK   Knicks   47  1306.72  1297.07      40.69    CHS   
4            0     DTF  Falcons   33  1300.00  1279.62      38.86    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68    1300.00    1306.72             H           L   
1   Huskies       66    1300.00    1293.28             A           W   
2    Knicks       47    1306.72    1297.07             H           W   
3     Stags       63    1300.00    1309.65             A           L   
4  Capitols       50    1300.00    1320.38             H           L   

   forecast notes  
0      0.64   NaN  
1      0.36   NaN  
2      0.63   NaN  
3      0.37   NaN  
4      0.64   NaN

In [0]:
pd.set_option("display.precision", 2)

In [44]:
nba.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts    elo_i    elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.00  1293.28      40.29    NYK   
1            0     NYK   Knicks   68  1300.00  1306.72      41.71    TRH   
2            0     CHS    Stags   63  1300.00  1309.65      42.01    NYK   
3            0     NYK   Knicks   47  1306.72  1297.07      40.69    CHS   
4            0     DTF  Falcons   33  1300.00  1279.62      38.86    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68    1300.00    1306.72             H           L   
1   Huskies       66    1300.00    1293.28             A           W   
2    Knicks       47    1306.72    1297.07             H           W   
3     Stags       63    1300.00    1309.65             A           L   
4  Capitols       50    1300.00    1320.38             H           L   

   forecast notes  
0      0.64   NaN  
1      0.36   NaN  
2      0.63   NaN  
3      0.37   NaN  
4      0.64   NaN

In [45]:
nba.tail(3)

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts    elo_i    elo_n  \
126311         101            1     CLE  Cavaliers   91  1704.39  1700.74   
126312         102            1     CLE  Cavaliers   97  1700.74  1692.09   
126313         103            1     GSW   Warriors  105  1813.63  1822.29   

        win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
126311      60.01    GSW   Warriors      104    1809.98    1813.63   
126312      59.29    GSW   Warriors      105    1813.63    1822.29   
126313      68.52    CLE  Cavaliers       97    1700.74    1692.09   

       game_location game_result  forecast notes  
126311             A           L      0.23   NaN  
126312             H           L      0.48   NaN  
126313             A           W      0.52   NaN

Use .info() to display all columns and data type

In [0]:
nba.info()

# Showing Basics Statistics

.describe() to see basic descriptive statistics for all numeric columns

In [0]:
nba.describe()

In [48]:
import numpy as np
nba.describe(include=np.object)

game_id   lg_id  date_game team_id fran_id  opp_id opp_fran  \
count         126314  126314     126314  126314  126314  126314   126314   
unique         63157       2      12426     104      53     104       53   
top     198812230MIL     NBA  4/16/2014     BOS  Lakers     BOS   Lakers   
freq               2  118016         30    5997    6024    5997     6024   

       game_location game_result           notes  
count         126314      126314            5424  
unique             3           2             231  
top                A           W  at New York NY  
freq           63138       63157             440

Take a look at the team_id and fran_id columns. Your dataset contains 104 different team IDs, but only 53 different franchise IDs. Furthermore, the most frequent team ID is BOS, but the most frequent franchise ID Lakers. How is that possible? You’ll need to explore your dataset a bit more to answer this question.


In [0]:
# Exploring Your Dataset

In [0]:
nba["team_id"].value_counts()

In [0]:
nba["fran_id"].value_counts()

DataFrame.loc¶
Access a group of rows and columns by label(s) or a boolean array.

.loc[] is primarily label based, but may also be used with a boolean array.

In [53]:
nba.loc[nba["fran_id"] == "Lakers", "team_id"].value_counts()

LAL    5078
MNL     946
Name: team_id, dtype: int64

In [55]:
nba.loc[nba['team_id'] == "MNL", "date_game"].min()

'1/1/1949'

In [56]:
nba.loc[nba["team_id"] == "MNL", "date_game"].max()

'4/9/1959'

DataFrame.agg(self, func, axis=0, *args, **kwargs)[source]
Aggregate using one or more operations over the specified axis.

In [58]:
nba.loc[nba["team_id"] == "MNL", "date_game"].agg(("min","max"))

min    1/1/1949
max    4/9/1959
Name: date_game, dtype: object

It looks like the Minneapolis Lakers played between the years of 1949 and 1959. That explains why you might not recognize this team!

###### Why the Boston Celtics team "BOS" played the most games in the dataset?
###### How many points the Boston Celtics have scored during all matches contained in this dataset?


In [62]:
nba.loc[nba["team_id"] == "BOS" , "pts"].agg("sum")

626484

# Getting to Know Pandas’ Data Structures

## Understanding Series Objects

Python’s most basic data structure is the list, which is also a good starting point for getting to know pandas.Series objects. Create a new Series object based on a list:



In [64]:
revenues = pd.Series([5555,7000,1990])
revenues

0    5555
1    7000
2    1990
dtype: int64

A Series object wraps two components:

A sequence of values
A sequence of identifiers, which is the index

In [70]:
revenues.values

array([5555, 7000, 1990])

In [69]:
revenues.index

RangeIndex(start=0, stop=3, step=1)

In [73]:
city_revenues = pd.Series([5555, 7000, 1990], index = ["NY","NJ","FL"])
city_revenues

NY    5555
NJ    7000
FL    1990
dtype: int64